pkg/ovsdb/monitor_test.go

```diff
@@ -152,98 +130,55 @@ func TestTableUpdate(t *testing.T) {
-		var expectedError error
-		dbServer := &DatabaseMock{
-			Response: expectedResponse,
-			Error:    expectedError,
-		}
-		ch := NewHandler(context.Background(), dbServer)
-		cmp := ovsjson.CondMonitorParameters{}
-		cmp.DatabaseName = dataBase
-		cmp.JsonValue = nil
-		mcr := ovsjson.MonitorCondRequest{}
-		for _, sel := range selects {
-			for _, col := range columns {
-				mcr.Columns = col
-				mcr.Select = sel
-				cmp.MonitorCondRequests = map[string][]ovsjson.MonitorCondRequest{tableName: {mcr}}
-
-				monitorResponse, actualError := ch.Monitor(nil, cmp)
-				assert.Nilf(t, actualError, "handler call should not return error: %v", err)
-
-				monitorCondResponse, actualError := ch.MonitorCond(nil, cmp)
-				assert.Nilf(t, actualError, "handler call should not return error: %v", err)
+func TestMonitor(t *testing.T) {
+	m := newMonitor("dbTest", &DatabaseMock{})
+	dbPrefix := "ovsdb"
+	mcr1 := ovsjson.MonitorCondRequest{Columns: []string{"c1","c3","c2"}}
+	mcr2 := ovsjson.MonitorCondRequest{Columns: []string{"c4"}}
+	mcr3 := ovsjson.MonitorCondRequest{Columns: []string{"a1"}}
+	u1 := mcrToUpdater(mcr1, true)
+	u2 := mcrToUpdater(mcr2, true)
+	u3 := mcrToUpdater(mcr3, true)
+	m1 := map[string][]*updater{"table1":{u1, u2}, "table2":{u3}}
+	h1 := handlerKey{jasonValue: "jasonValue1"}
+	m.addUpdaters(dbPrefix, m1, h1)
+	assert.Equal(t, map[handlerKey]bool{h1:true}, m.handlers,"Handlers maps should be equals")
+	assert.Equal(t, map[string][]updater{"ovsdb/dbTest/table1":{*u1,*u2}, "ovsdb/dbTest/table2":{*u3}}, m.key2Updaters,"Key to updater maps should be equals")
+	assert.Equal(t, map[string][]handlerKey{u1.key:{h1}, u2.key:{h1}, u3.key:{h1}}, m.upater2handlers)
+	assert.Truef(t, m.hasHandlers(), "Monitor should have updaters")
+
+	h2 := handlerKey{jasonValue: "jasonValue2"}
+	m.addUpdaters(dbPrefix, m1, h2)
+	assert.Equal(t, map[handlerKey]bool{h1:true, h2:true}, m.handlers,"Handlers maps should be equals")
+	assert.Equal(t, map[string][]updater{"ovsdb/dbTest/table1":{*u1,*u2}, "ovsdb/dbTest/table2":{*u3}}, m.key2Updaters,"Key to updater maps should be equals")
+	assert.Equal(t, map[string][]handlerKey{u1.key:{h1, h2}, u2.key:{h1, h2}, u3.key:{h1, h2}}, m.upater2handlers)
+
+	u11 := mcrToUpdater(mcr1, false)
+	m11 := map[string][]*updater{"table1":{u11}}
+	h11 := handlerKey{jasonValue: "jasonValue11"}
+	m.addUpdaters(dbPrefix, m11, h11)
+	assert.Equal(t, map[handlerKey]bool{h1:true, h2:true, h11:true}, m.handlers,"Handlers maps should be equals")
+	assert.Equal(t, map[string][]updater{"ovsdb/dbTest/table1":{*u1,*u2, *u11}, "ovsdb/dbTest/table2":{*u3}}, m.key2Updaters,"Key to updater maps should be equals")
+	assert.Equal(t, map[string][]handlerKey{u1.key:{h1, h2}, u2.key:{h1, h2}, u3.key:{h1, h2}, u11.key:{h11}}, m.upater2handlers)
+
+	m.removeUpdaters(dbPrefix, map[string][]string{"table1":{u11.key}}, h11)
+	assert.Equal(t, map[handlerKey]bool{h1:true, h2:true}, m.handlers,"Handlers maps should be equals")
```

Comment:
you are checking against the internal state in `m.handlers` this makes you tests verify sensitive to changes in implementation, and alternative would be to expose semantics which enable to get state externally which would enable you to implement the tests

Reply:
I agree about the internal state, and sensitivity of the tests, but the internal state is quite complicated, and I want to be sure that it is updated correctly.